spectator_hud: add widget spectator hud

keelefi opened the pull request:
<!--
PR Template! Please make sure to give your PR a relevant title so a squash merge remains descriptive
If any commented sections are not relevant to this PR, remove them.
Please fill out the uncommented sections with any relevant information.
-->

### Work done
<!--
Describe the changes or additions made in this PR, and why they
are necessary or important. If there is unusual complexity in the
code or functionality, please explain it so reviewers can understand.
-->
Add a new widget Spectator HUD. This widget has been in development for some time and the relevant Discord there is here: https://discord.com/channels/549281623154229250/1182353569576407060

#### Test steps
- [ ] Launch game in spectator mode
- [ ] Watch bots build bases and fight and notice the metrics displayed by Spectator HUD change as the game state changes

### Screenshots:

#### BEFORE:
![pr-screenshot-before](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/d5c24c98-9c58-4a24-8c3c-01b63d7520d7)


#### AFTER:
![pr-screenshot-after](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/64ebc2a7-c365-44e8-9f1d-608c8f933f05)

salinecitrine reviewed (commented):
First of all, this is an awesome widget; I use it all the time, and I've seen lots of others use it too :)

I did a skim through the code, and added some quick comments. I'll try to take a deeper look later.

Some more general notes:

* Performance is pretty bad, to the point that it's using 1.5+ orders of magnitude more time and memory that any other widget. I think that in order to get this merged, that will have to be addressed. That said, I'm not sure if we have specific guidelines to aim for.
* The standard formatting for lua code in this repo is 4-width tabs, while this file uses spaces (See https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/.editorconfig)
* You have trailing whitespaces on a few lines. I recommend using something to automatically format your code on save (an IDE, probably)

keelefi commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+local headerDimensions = {}
+
+local topBarPosition
+local topBarShowButtons
```
Fixed in a7b40d6c1

keelefi commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+end
+
+local function addToUnitCache(teamID, unitID, unitDefID)
+    local function addToUnitCacheInternal(cache, teamID, unitID, value)
```
Fixed in ff3c36a72

nvm this didn't work. Reverted in 5525cc909. Without the revert widget fails:

```
Error in Initialize(): [string "LuaUI/Widgets/gui_spectator_hud.lua"]:453: attempt to index local 'value' (a number value)
```

salinecitrine replied:
You'll also need to update any locations where it's called to remove those parameters as well. The error you're getting is because it expects the second argument to be `value`, but it's getting `teamID`, so all of the values end up set to the `teamID` instead.

keelefi commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+local function buildMetricsEnabled()
+    metricsEnabled = {}
+
+    index = 1
```
Fixed in 66e89c684

keelefi commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+            headerDimensions.left,
+            highlightBottom,
+            headerDimensions.right,
+            highlighTop
```
Fixed in b95707f68

keelefi commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+        local statsRight = vsModeStats[indexRight].metrics[metric.id]
+
+        local bottom = vsModeMetricsAreaTop - metric.id * vsModeMetricHeight
+        local top = bottom + vsModeMetricHeight
```
Fixed in ebe0814e1

keelefi commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+    if ((valueLeft > 0) or (valueRight > 0)) and
+            ((mouseX > left) and (mouseX < right) and (mouseY > bottom) and (mouseY < top)) then
+        local scalingFactor = barLength / (valueLeft + valueRight)
+        local lineMiddle = mathfloor((top + bottom) / 2)
```
Fixed in cf457c3f5

keelefi commented:
>     * The standard formatting for lua code in this repo is 4-width tabs, while this file uses spaces (See https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/.editorconfig)

Fixed in 099c29ba5

>     * You have trailing whitespaces on a few lines.

Fixed in c169135be

>     * Performance is pretty bad, to the point that it's using 1.5+ orders of magnitude more time and memory that any other widget (as per the Widget Profiler widget). I think that in order to get this merged, that will have to be addressed. That said, I'm not sure if we have specific guidelines to aim for.

Yeah, I have used lots of time and effort trying to chase performance problems. My competence in Lua and OpenGL ends here and would need help to improve further.

salinecitrine commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2307 @@
+        end
+    end
+
+    local function addSpaces(number)
```
Ah, I see, I didn't notice that it references itself. In that case I would move `addSpaces` outside of `formatResources` so it's not getting redefined on every function call.

thehobojoe commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2288 @@
+	return mathfloor(num * mult + 0.5) / mult
+end
+
+local function formatResources(amount, short)
```
This can be replaced with the new `formatSI` functionality from @salinecitrine

keelefi replied:
I'm in the camp that for different UI elements, different formatting makes sense. We have one style in top bar, other style in adv playerlist and a third style in stats panel. In my humble opinion, it's not an oversight, but rather a feature.

thehobojoe commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2288 @@
+local function getOneStat(statKey, teamID)
+	local result = 0
+
+	if statKey == "metalIncome" then
```
Not critical, but for safety from typos and to allow autocomplete, it would be good if these keys were stored in variables in a table. Like
```
keys = {
metalIncome = "metalIncome",
}
```
etc, which can then be accessed with `keys.metalIncome`

keelefi replied:
Done.

thehobojoe commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,2288 @@
+		headerDimensions.bottom
+	)
+
+	-- TODO: this is not working, find out why
+	local mouseX, mouseY = spGetMouseState()
+	if (mouseX > headerDimensions.left) and
+			(mouseX < headerDimensions.right) and
+			(mouseY > headerDimensions.bottom) and
+			(mouseY < metricChangeBottom) then
```
What's not working here?

Mouse position checks like this can be simplified a fair bit by using a construct like `Rect` from gridmenu https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_gridmenu.lua#L154

This can be used for nice concise checks just going `header:contains(mouseX, mouseY)`. Again not a critical thing, and maybe not worth doing until I abstract that construct somewhere accessible by other widgets, but I haven't really gotten around to it and it may never happen with RMLUI around the corner.

thehobojoe reviewed (commented):
Overall, very nice. The code is exceptionally well organized and formatted which makes it very pleasant to read, great work there.

My only qualm about merging it is the performance is *quite* bad, easily tripling the performance cost of advplayerlist which is the current most expensive widget. I think the default update rate should be lowered at a minimum, and perhaps reactively reduce itself if sim speed slows down, or if the player is catching up in the simulation.

keelefi commented:
Thank you @thehobojoe for taking the time and effort to review this PR. Much appreciated.

> My only qualm about merging it is the performance is _quite_ bad

Yes, you are absolutely correct that there are some concerning performance issues with Spectator HUD. However, the whole story is a bit more complicated. Allow me to share my findings.

My typical test setup is to load up a skirmish on Quicksilver with a few AI's on each side. In this scenario, there is not much going on for the first few minutes. Hence, the FPS is very high. For example, in this screenshot I have 171 FPS:

![screen_2024-03-27_14-37-17-228](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/f4807c75-ebc3-4d49-9688-4ebb10b4a385)

Looking at the readings from Widget Profiler, Spectator HUD is quite a perpetrator. At 6% of CPU usage, Spectator HUD has the biggest reading whereas Top Bar at 2% comes in second:

![screen_2024-03-27_14-37-33-968](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/dc3fd5b2-3325-4d33-9a8c-63f5c8842833)

However, looking more carefully at the readings we notice that most CPU time is spent in `DrawScreen()`. Remember that `DrawScreen()` is called on each frame. Therefore, the higher the FPS, `DrawScreen()` is called more frequently.

Opening Widget Selector will drop FPS massively. In my case, from 171 FPS to less then 100 FPS:

![screen_2024-03-27_14-37-52-079](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/ffbc88ae-b2b8-42a2-9782-1fde901c0e40)

We see that Widget Selector is using nearly 70% of CPU usage and Spectator HUD has dropped to 3%.

In the following screenshot Widget Selector is disabled and game is resumes (no longer paused). In this scenario, Top Bar starts using a lot more CPU. Spectator HUD sits at 5% with Top Bar matching at around 5% CPU usage:

![screen_2024-03-27_14-38-10-988](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/dfa42afc-8076-4487-8345-e8921f4fff44)

Indeed, in scenarios with less FPS Spectator HUD CPU usage drops because `DrawScreen()` is called less frequently. Furthermore, many other widgets start using more CPU by comparison, further dropping the relative share of Spectator HUD.

Here's a late-game catch-up (~4x simulation speed) screenshot:

![screen_2024-03-27_14-55-22-711](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/a324f8f2-031f-4f19-aa4f-013406f17a30)

Spectator HUD CPU usage sits now at 2% matching that of AdvPlayerList. The biggest perpetrator now is Top Bar at over 12% CPU usage.

Also noteworthy in this scenario is that the biggest CPU usage of Spectator HUD has now switched from `DrawScreen()` to `GameFrame()`.

A screenshot from the same late-game, but now without catch-up (1.0x simulation speed):

![screen_2024-03-27_14-55-50-372](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/11913019/ba63a7ee-e1b9-4bb6-9dbe-84d67947c0ac)

Spectator HUD is now around 3%.

As I admitted earlier, there are performance concerns with Spectator HUD. Only the fact that Widget Profiler shows nasty readings during early game is enough reason for me to be willing to fix the problems.

Unfortunately, I have already spent an unreasonable amount of time and effort chasing Spectator HUD performance problems. Every recommended Lua performance optimization has been implemented. If the remaining problems are related to Lua or OpenGL, I cannot tell. My skills in both are lacking.

If it was the case that by spending 10 more hours debugging, the problems would be fixed, I would be willing to invest that time. Sadly, without better tools or help from others, it would likely be 10 more hours wasted without real performance gains.

> I think the default update rate should be lowered at a minimum, and perhaps reactively reduce itself if sim speed slows down, or if the player is catching up in the simulation.

This is unlikely to result in any performance gains. The collection of statistics done in `GameFrame()` is hardly a performance problem.

Instead, if it was possible to render the widget graphics into a buffer in `DrawScreen()` less frequently, that might help. Currently, all the graphics of the widget are re-rendered for every frame. But this makes little sense as the data displayed changes only 15 times per second (at most). So, I'm wondering if we could render the data 15 times per second and for the rest of the frames we would just use data from the buffer.

But I know so little OpenGL I don't know if this is doable or if it's even the right thing to do. It might very well be OpenGL has some more sophisticated and less NIH solutions for this problem. Maybe it's DisplayLists, maybe the widget is using shaders wrong...

So, how to progress in this area? If there's a tool that can create flame graphs for individual widgets that would be great. Then we could maybe get more accurate data as to where the problem or problems are.

thehobojoe commented:
> Unfortunately, I have already spent an unreasonable amount of time and effort chasing Spectator HUD performance problems. Every recommended Lua performance optimization has been implemented. If the remaining problems are related to Lua or OpenGL, I cannot tell. My skills in both are lacking.

First, don't be down on yourself, this is a very nice widget, and optimization of this nature is very difficult. 🙂 

Optimizing draw performance is very complicated, and I'd need to spend more time studying this widget to give any meaningful recommendations, but I suspect it's not the main problem. Accurate profiling is really difficult, to do it comprehensively you'd definitely need to use Tracy, but you can do quite a lot with just checking how long something takes to run with `os.clock()`. How much time is spent in `updateStats` is probably the primary thing in larger games.

But most importantly, the default setting causing 15 updates per second is *WAY* too high. The information this widget displays does not need to update even remotely that quickly to give accurate and relevant information, even twice per second is probably too much. Sure, it can look cool to have the sliders moving around really reactively, but it doesn't actually explain anything useful and is not worth the performance cost.

Hopefully with RMLUI the draw performance part can be improved a lot, but some of it may be fixable without that.

RebelNode commented:
Tested it, seems to work as advertised.

I would probably put "enabled = true" by default if it gets merged into main game, not much of a reason to hide it behind settings if it's decided to be good for the game.

Beherith commented on luaui/Widgets/gui_spectator_hud.lua:
```diff
@@ -0,0 +1,1949 @@
+	end
+
+	if movingAverage.index == 0 then
+		for i=1,settings.movingAverageWindowSize do
```
For the record, you dont actually need recalc this with a loop:
newavg = (oldavg * (settings.movingAverageWindowSize -1 ) + newvalue) / settings.movingAverageWindowSize

keelefi replied:
Oh boy, this is implemented so long ago that I no longer remember the exact reason for it :/ This should definitely have a comment explaining what's going on because it's far from obvious. Allow me to explain the code.

In Lua, indexes start from 1. The moving average code here initializes the index to 0 which is a special value. When updating the moving average for the first time with a value, all of the window is set to the value (i.e. the first observed value).

The normal case where we just update a single value has no looping. We just replace the oldest value with the newest value. See line 816 with the code:

```
	movingAverage.average = movingAverage.average + (newValue - oldValue) / settings.movingAverageWindowSize
```

I think the reason for populating the window with the first encountered value is because of the scenario where the widget is enabled mid game. If the window was initialized with all zero's, the widget would show increasing stats for everything until the moving average window is filled with non-bogus values.

Beherith reviewed (commented):
Could be useful to check out the load with Tracy

Beherith commented:
Ill try to put it through the grinder tracy tomorrow

Beherith commented:
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/109391/98fa61e0-ee3d-45c3-ab13-8be15118137e)

Beherith commented:
Doesnt look all that bad:
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/109391/98f8f507-a65e-4ba0-9e75-268f32ce8fe9)

Beherith commented:
![image](https://github.com/beyond-all-reason/Beyond-All-Reason/assets/109391/2afbbe23-7e2b-433d-8d27-03e98e9116a2)

Beherith commented:
So I think its OK on perf, altogether.

Beherith commented:
The REAL nasty perf tricks start when you begin to use render-to-texture to speed up static text and static UI element drawing, See AtlasOnDemand library. 
But I applaud you for using VBO's to begin with!

Beherith commented:
The Mouseover is neato!

keelefi commented:
> Render-to-texture (texture buffers) are a pain in the ass to do with text.

Thank you very much @Beherith !

I added `RenderToTexture` in https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2607/commits/392a4bde28aae6edb0798b264acc3ca69a718598 and the perf is now a lot better!

Damgam commented:
@keelefi Is this ready to merge? I'm gonna push the button if so.

keelefi commented:
> @keelefi Is this ready to merge? I'm gonna push the button if so.

Yes, please. Thank you.
